File: 프로그래머스/unrated/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int solution(vector<vector<int>> land) {
    int answer = 0;
    vector<int> visit[land.size()];
    vector<int> oil[land.size()];
    queue<pair<int,int>> q;
    int number =0;
    vector<int> oil_size;
    int dir[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};
    for(int i=0;i<land.size();i++){
        for(int j=0;j<land[i].size();j++){
            visit[i].push_back(0);
            oil[i].push_back(-1);
        }
    }
    for(int i=0;i<land.size();i++){
        for(int j=0;j<land[i].size();j++){
            if(land[i][j]==1&&visit[i][j]==0){
                q.push({i,j});
                visit[i][j]=1;
                int total=1;
                while(!q.empty()){
                    int x = q.front().first;
                    int y = q.front().second;
                    //visit[i][j]=1;
                    oil[x][y]=number;
                    q.pop();
                    for(int d=0; d < 4; d++){
                        int a = x+dir[d][0];
                        int b = y+dir[d][1];
                        if(a<0||a>=land.size()||b<0||b>=land[d].size())continue;
                        if(visit[a][b]==1||land[a][b]==0)continue;
                        q.push({a,b});
                        visit[a][b]=1;
                        total++;
                    }
                }
                oil_size.push_back(total);
                number++;
            }
        }
    }
    for(int j=0;j<land[0].size();j++){
        set<int> s;
        int temp_size=0;
        for(int i=0;i<land.size();i++){
            if(oil[i][j]>-1){
                if(s.count(oil[i][j])==0){
                    s.insert(oil[i][j]);
                    temp_size+=oil_size[oil[i][j]];
                }
            }
        }
        answer = max(answer,temp_size);
    }
    return answer;
}

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.js
function solution(land) {
    var answer = 0;
    const n = land.length;
    const m = land[0].length;
    const dir = [[1,0],[-1,0],[0,1],[0,-1]]
    const value = new Map();
    const visited = Array.from({length:land.length},()=>{
        return Array.from({length:land[0].length},()=>false)
    })
    const oil_map = Array.from({length:m},()=>{
        return new Set();
    })
    let oil_num = 2;
    for(let i=0;i<n;i++){
        for(let j=0;j<m;j++){
            const oil = [[i,j]];
            let oil_amount = 0;
            while(oil.length>0){
                const [a,b] = oil.shift();
                if(land[a][b]===1&&!visited[a][b]){
                    land[a][b] = oil_num
                    visited[a][b] = true;
                    oil_map[b].add(oil_num)
                    oil_amount++;
                    for(let d=0;d<4;d++){
                        const x = a+dir[d][0];
                        const y = b+dir[d][1];
                        if(x<0||x>=n||y<0||y>=m) continue;
                        if(visited[x][y]||land[x][y]===0) continue;
                        oil.push([x,y])
                    }
                }
            }
            if(oil_amount>0){
                value[oil_num] = oil_amount
                oil_num++
            }
        }
    }
    for(let i=0;i<m;i++){
        let result = 0;
        for(const o of oil_map[i]){
            result+=value[o]
        }
        answer = Math.max(result,answer)
    }
    return answer;
}